src/ast/nodes/shared/Node.ts

```diff
@@ -143,7 +159,16 @@ export class NodeBase implements ExpressionNode {
 	}
 
 	hasEffects(options: ExecutionPathOptions): boolean {
-		return this.someChild((child: NodeBase) => child.hasEffects(options));
+		for (const key of this.keys) {
+			const value = (<GenericEsTreeNode>this)[key];
+			if (value === null) continue;
+			if (Array.isArray(value)) {
+				for (const child of value) {
+					if (child !== null && child.hasEffects(options)) return true;
+				}
+			} else if (value.hasEffects(options)) return true;
+		}
+		return false;
```

Comment:
I know this is one of the ugliest parts of this refactoring as we basically have the same loop four times with only slight variations (five times if you count `parseNode` but this method also does a lot more). The performance gains, however, were pretty substantial. Of the 50% less tree-shaking time, *at least* 15-20% can be accounted to these loops.

Unless we change everything fundamentally (and I thought about this many times already; did not come up with a good approach yet which I would expect to provide an actual performance-*improvement* due to the zoo of different node effects—run, called, accessed, assigned + circularity prevention mechanisms for each—that need to be taken into account), I think this is close to the best we can do. Even a magically better optimized visitor capable of execting arbitrary methods on nodes would probably still be callback based and thus need two nested function calls to e.g. call `bind` on a child + the additional call to the visitor itself. I think this is as flat as it can get in the current pattern. Incidentally, call stacks are also more than halved with this approach which is not a paradigm shift but at least should not hurt.

As all loops of this kind are in `Node`, it should not be difficult to change them together if we find a better approach in the future.

Reply:
Yes the stack savings alone seem worth inlining these loops I suppose, even if it means forgoing the visitor pattern.

Btw I did experiment with a flat for-based visitor pattern on a performance branch previously to avoid stack calls and there was an absolutely negligible performance change, so that might even void the argument that a unified visitor could be more optimized.